ktlint-ruleset-standard/src/main/kotlin/com/github/shyiko/ktlint/ruleset/standard/IndentationRule.kt

```diff
@@ -2,81 +2,312 @@ package com.github.shyiko.ktlint.ruleset.standard
-            indentSize = gcd(maxOf(ec.indentSize, 1), maxOf(ec.continuationIndentSize, 1))
-            return
-        }
+        val ec = EditorConfig.from(node as FileASTNode)
+        indentSize = ec.indentSize
+        continuationIndentSize = ec.continuationIndentSize
+
         if (indentSize <= 1) {
             return
         }
-        if (node is PsiWhiteSpace && !node.isPartOf(PsiComment::class)) {
-            val lines = node.getText().split("\n")
-            if (lines.size > 1 && !node.isPartOf(KtTypeConstraintList::class)) {
-                var offset = node.startOffset + lines.first().length + 1
-                val previousIndentSize = node.previousIndentSize()
-                lines.tail().forEach { indent ->
-                    if (indent.isNotEmpty() && (indent.length - previousIndentSize) % indentSize != 0) {
-                        if (!node.isPartOf(KtParameterList::class)) { // parameter list wrapping enforced by ParameterListWrappingRule
-                            emit(
-                                offset,
-                                "Unexpected indentation (${indent.length}) (it should be ${previousIndentSize + indentSize})",
-                                false
-                            )
-                        }
+
+        exploreTree(node, emit, autoCorrect)
+    }
+
+    private fun exploreTree(
+        node: ASTNode,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit,
+        autoCorrect: Boolean
+    ) {
+        val scopeStack = Stack<IndentationScope>()
+        val stack = Stack<ASTNode>()
+
+        val scopeSet = setOf(
+            IndentationScope.ExpressionScope(
+                listOf(KtStubElementTypes.DOT_QUALIFIED_EXPRESSION, KtNodeTypes.SAFE_ACCESS_EXPRESSION),
+                listOf(KtTokens.DOT, KtTokens.SAFE_ACCESS)
+            ),
+            IndentationScope.ExpressionScope(
+                listOf(KtNodeTypes.BINARY_EXPRESSION),
+                listOf(KtNodeTypes.OPERATION_REFERENCE)
+            ),
+            IndentationScope.InternalNodeScopeWithTriggerToken(
+                KtStubElementTypes.PROPERTY,
+                KtTokens.EQ
+            ),
+            IndentationScope.InternalNodeScopeWithTriggerToken(
+                KtStubElementTypes.FUNCTION,
+                KtTokens.EQ
+            ),
+            IndentationScope.InternalNodeScopeWithTriggerToken(
+                KtNodeTypes.WHEN_ENTRY,
+                KtTokens.ARROW
+            ),
+            IndentationScope.ParenScope(KtTokens.LBRACE, KtTokens.RBRACE),
+            IndentationScope.ParenScope(KtTokens.LBRACKET, KtTokens.RBRACKET),
+            IndentationScope.ParenScope(KtTokens.LPAR, KtTokens.RPAR),
+            IndentationScope.ParenScope(KtTokens.LT, KtTokens.GT),
+            IndentationScope.PropertyAccessorScope
+        )
+
+        stack.push(node)
+
+        while (!stack.empty()) {
```

Comment:
Kotlin standard library also offers [`isNotEmpty()`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/is-not-empty.html).